src/hubbleds/pages/05-class-results-&-uncertainty/__init__.py

```diff
@@ -1,535 +1,709 @@
+        if not value:
+            return
+        
+        class_ids = LOCAL_STATE.value.stage_5_class_data_students
+        class_data_points = [m for m in LOCAL_STATE.value.class_measurements if m.student_id in class_ids]
+        class_data = models_to_glue_data(class_data_points, label="Class Data")
+        class_data = GLOBAL_STATE.value.add_or_update_data(class_data)
+
+        layer_viewer = viewers["layer"]
+        layer_viewer.state.x_axislabel = "Distance (Mpc)"
+        layer_viewer.state.y_axislabel = "Velocity"
+        layer_viewer.add_data(class_data)
+        layer_viewer.state.x_att = class_data.id['est_dist_value']
+        layer_viewer.state.y_att = class_data.id['velocity_value']
+        class_data_added.set(True)
+
+        if len(class_data.subsets) == 0:
+            student_slider_subset = class_data.new_subset(label="student_slider_subset", alpha=1, markersize=10)
+        else:
+            student_slider_subset = class_data.subsets[0]
+        slider_viewer = viewers["student_slider"]
+        slider_viewer.add_data(class_data)
+        slider_viewer.state.x_att = class_data.id['est_dist_value']
+        slider_viewer.state.y_att = class_data.id['velocity_value']
+        slider_viewer.state.title = "Stage 5 Class Data Viewer"
+        slider_viewer.add_subset(student_slider_subset)
+        slider_viewer.layers[0].state.visible = False
+
+        class_summary_data = make_summary_data(class_data,
+                                               input_id_field="student_id",
+                                               output_id_field="id",
+                                               label="Class Summaries")
+        class_summary_data = GLOBAL_STATE.value.add_or_update_data(class_summary_data)
+
+        hist_viewer = viewers["student_hist"]
+        hist_viewer.add_data(class_summary_data)
+        hist_viewer.state.x_att = class_summary_data.id['age_value']
+        hist_viewer.state.title = "My class ages (5 galaxies each)"
+        hist_viewer.layers[0].state.color = "red"
+
+    class_data_loaded.subscribe(_on_class_data_loaded)
+
+    measurements_loaded = Ref(LOCAL_STATE.fields.measurements_loaded)
+    student_data_added = solara.use_reactive(False)
+    def _on_student_data_loaded(value: bool):
+        if not value:
+            return
+        student_data = models_to_glue_data(LOCAL_STATE.value.measurements, label="My Data", ignore_components=["galaxy"])
+        # NB: If there are no components, Data::size returns 1 (empty product)
+        # so that can't be our check
+        if not student_data.components:
+            student_data = empty_data_from_model_class(StudentMeasurement, label="My Data")
+        student_data = GLOBAL_STATE.value.add_or_update_data(student_data)
+        student_data_added.set(True)
+
+
+    if measurements_loaded.value:
+        _on_student_data_loaded(True)
+    else:
+        measurements_loaded.subscribe(_on_student_data_loaded)
```

Comment:
This is a guard in case the async loading of the measurements isn't completed by the time this page renders?

Reply:
Yeah, that's the idea. The measurements will probably have been loaded in a previous step, but that won't be the case if this is the first page that the student is coming to.